### PR TITLE
Ebs ta windows bug fixes

### PR DESCRIPTION
**Pull request by saurabhc123**
### Summary

This PR added some bug fixes to the code base in the form of adding some additional duration for Windows specific staging and unstaging timeouts.

### Implementation details

We refactored the existing constants for windows and linux separately and specified the platform specific values separately.

### Testing

UTs used for Linux was repurposed for Windows and are all passing.

New tests cover the changes:  There are no new tests for timeouts as the code was just refactored. The other tests were repurposed and refactored.

### Description for the changelog

Add custom EBS Task-attach timeout durations for Windows

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by mye956 — COMMENTED** [review]
When you get a chance, can you call out what will be changed, a couple of implementation details (high-level), and how you've tested the changes within the PR description?

File: agent/engine/task_manager.go [hunk 1]
```diff
@@ -61,7 +60,7 @@ const (
 	maxStoppedWaitTimes                      = 72 * time.Hour / stoppedSentWaitInterval
 	taskUnableToTransitionToStoppedReason    = "TaskStateError: Agent could not progress task's state to stopped"
 	// unstage retries are ultimately limited by successful unstage or by the unstageVolumeTimeout
-	unstageVolumeTimeout = 30 * time.Second
+	unstageVolumeTimeout = 600 * time.Second
```
Comment (mye956) [inline agent/engine/task_manager.go:63]: This is a pretty big jump in the unstage timeout duration. I know that some processes can vary between OS types/platforms, could this be scoped down for just windows without affecting for all?

File: ecs-agent/daemonimages/csidriver/driver/node.go [hunk 2]
```diff
@@ -69,7 +70,8 @@ type nodeService struct {
 }
 
 func (d *nodeService) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
-	klog.V(4).InfoS("NodeStageVolume: called", "args", *req)
```
Comment (mye956) [inline ecs-agent/daemonimages/csidriver/driver/node.go:72]: Q: Why are we leaving out `V(4)`?

File: ecs-agent/daemonimages/csidriver/driver/node.go [hunk 3]
```diff
@@ -262,7 +291,8 @@ func newNodeService() nodeService {
 }
 
 func (d *nodeService) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstageVolumeRequest) (*csi.NodeUnstageVolumeResponse, error) {
-	klog.V(4).InfoS("NodeUnstageVolume: called", "args", *req)
+	klog.InfoS("NodeUnstageVolume: called", "args", *req)
+	t := time.Now()
```
Comment (mye956) [inline ecs-agent/daemonimages/csidriver/driver/node.go:295]: nit: could you make the variable name a bit more descriptive?

File: ecs-agent/daemonimages/csidriver/driver/node.go [hunk 4]
```diff
@@ -69,7 +70,8 @@ type nodeService struct {
 }
 
 func (d *nodeService) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
-	klog.V(4).InfoS("NodeStageVolume: called", "args", *req)
+	klog.InfoS("NodeStageVolume: called", "args", *req)
+	t := time.Now()
```
Comment (mye956) [inline ecs-agent/daemonimages/csidriver/driver/node.go:74]: nit: could you make the variable name a bit more descriptive?

File: ecs-agent/daemonimages/csidriver/driver/node.go [hunk 5]
```diff
@@ -80,6 +82,10 @@ func (d *nodeService) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
 	if len(target) == 0 {
 		return nil, status.Error(codes.InvalidArgument, "Staging target not provided")
 	}
+	targetSplice := strings.Split(target, "\\")
```
Comment (mye956) [inline ecs-agent/daemonimages/csidriver/driver/node.go:85]: nit: could we add a brief comment on what we're trying to get here and what it's being used (e.g. this is being used for logging purposes)

File: agent/engine/task_manager.go (hunk 1 above)
Comment (saurabhc123) [reply agent/engine/task_manager.go:63]: Yes, we will refactor this to just affect windows and not linux.

File: ecs-agent/daemonimages/csidriver/driver/node.go (hunk 2 above)
Comment (saurabhc123) [reply ecs-agent/daemonimages/csidriver/driver/node.go:72]: We added the V(4) to all logging statements.

File: ecs-agent/daemonimages/csidriver/driver/node.go (hunk 3 above)
Comment (saurabhc123) [reply ecs-agent/daemonimages/csidriver/driver/node.go:295]: Done

File: ecs-agent/daemonimages/csidriver/driver/node.go (hunk 4 above)
Comment (saurabhc123) [reply ecs-agent/daemonimages/csidriver/driver/node.go:74]: Done

File: ecs-agent/daemonimages/csidriver/driver/node.go (hunk 5 above)
Comment (saurabhc123) [reply ecs-agent/daemonimages/csidriver/driver/node.go:85]: Done

**Review by mye956 — DISMISSED** [review]

File: agent/engine/task_manager.go [hunk 6]
```diff
@@ -1618,6 +1615,12 @@ func (mtask *managedTask) unstageVolumeWithRetriesAndTimeout(csiClient csiclient
 	derivedCtx, cancel := context.WithTimeout(mtask.ctx, unstageVolumeTimeout)
 	defer cancel()
 	backoff := retry.NewExponentialBackoff(unstageBackoffMin, unstageBackoffMax, unstageBackoffJitter, unstageBackoffMultiple)
+	now := time.Now()
+	defer func() {
+		logger.Debug("NodeUnstageVolume duration stats:", logger.Fields{
+			"duration": strconv.FormatFloat(time.Since(now).Seconds(), 'f', -1, 64),
```
Comment (singholt) [inline agent/engine/task_manager.go:1621]: why use strconv.FormatFloat here but not above in watcher.go?

File: agent/engine/task_manager_linux.go [hunk 7]
```diff
@@ -0,0 +1,22 @@
+//go:build linux
```
Comment (singholt) [inline agent/engine/task_manager_linux.go:1]: like other os specific logic in the agent repository, we need a _unsupported.go file for this

File: ecs-agent/daemonimages/csidriver/driver/node.go [hunk 8]
```diff
@@ -81,6 +83,13 @@ func (d *nodeService) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
 		return nil, status.Error(codes.InvalidArgument, "Staging target not provided")
 	}
 
+	// The following code parses the taskID from the given mount path. The target mount path is always in the
+	// format of C:\\ProgramData\\Amazon\\ECS\\ebs\\taskId-volumeId
```
Comment (singholt) [inline ecs-agent/daemonimages/csidriver/driver/node.go:87]: sounds like this needs to go in a Windows specific file and in its own helper function

File: ecs-agent/daemonimages/csidriver/driver/node.go [hunk 9]
```diff
@@ -154,19 +163,26 @@ func (d *nodeService) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
...
+	klog.V(4).Infof(" NodeStageVolume: findDevicePath succeeded for task %s and volume %s, time taken: %f seconds", taskId, volumeID, time.Since(findevicepathTime).Seconds())
 
+	pathexiststime := time.Now()
 	exists, err := d.mounter.PathExists(target)
-	klog.InfoS("NodeStageVolume: path exists:", "exists", exists)
+	klog.V(4).Infof("NodeStageVolume: PathExists for task %s and volume %s, time taken: %f seconds", taskId, volumeID, time.Since(pathexiststime).Seconds())
+	klog.V(4).InfoS("NodeStageVolume: path exists:", "exists", exists)
 	if err != nil {
-		klog.InfoS("NodeStageVolume: path exists:", "err", err)
+		klog.V(4).InfoS("NodeStageVolume: path exists:", "err", err)
```
Comment (singholt) [inline ecs-agent/daemonimages/csidriver/driver/node.go:185]: should this "path doesn't exist", if there's a err?

File: ecs-agent/daemonimages/csidriver/driver/node.go [hunk 10]
```diff
@@ -154,19 +163,26 @@ func (d *nodeService) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
 		if part != "0" {
 			partition = part
 		} else {
-			klog.InfoS("NodeStageVolume: invalid partition config, will ignore.", "partition", part)
+			klog.V(4).InfoS("NodeStageVolume: invalid partition config, will ignore.", "partition", part)
 		}
 	}
 
+	klog.V(4).Infof("NodeStageVolume: configs initialized for task %s and volume %s, time taken: %f seconds", taskId, volumeID, time.Since(nodeStageStartTime).Seconds())
```
Comment (singholt) [inline ecs-agent/daemonimages/csidriver/driver/node.go:170]: General question about the log statement additions in this file:

What does logging some of these time measurements in this file help us with? 
Are there retries within each methods (findDevicePath, PathExists, etc.) that may take longer and help the operator in debugging?

there's quite a bit of logging added, and all of them are info-level logs and per EBS volume attachment, perhaps making the logs too noisy?

File: ecs-agent/daemonimages/csidriver/driver/node.go [hunk 11]
```diff
@@ -154,19 +163,26 @@ func (d *nodeService) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
...
+
+	findevicepathTime := time.Now()
```
Comment (singholt) [inline ecs-agent/daemonimages/csidriver/driver/node.go:172]: please use camel case for this and all time.Now() additions below.

File: ecs-agent/daemonimages/csidriver/driver/node.go [hunk 12]
```diff
@@ -175,16 +191,21 @@ func (d *nodeService) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
 	// Otherwise we need to create the target directory.
 	if !exists {
 		// If target path does not exist we need to create the directory where volume will be staged
-		klog.InfoS("NodeStageVolume: creating target dir", "target", target)
+		makedirtime := time.Now()
```
Comment (singholt) [inline ecs-agent/daemonimages/csidriver/driver/node.go:194]: related to my overall comment above, why is this useful?

File: ecs-agent/daemonimages/csidriver/driver/node.go [hunk 13]
```diff
@@ -273,6 +306,13 @@ func (d *nodeService) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
 		return nil, status.Error(codes.InvalidArgument, "Staging target not provided")
 	}
 
+	// The following code parses the taskID from the given mount path. The target mount path is always in the
+	// format of C:\\ProgramData\\Amazon\\ECS\\ebs\\taskId-volId
```
Comment (singholt) [inline ecs-agent/daemonimages/csidriver/driver/node.go:310]: Windows-specific code?

File: agent/engine/task_manager.go (hunk 6 above)
Comment (singholt) [reply agent/engine/task_manager.go:1621]: Seconds() returns a float: https://pkg.go.dev/time#Duration.Seconds

File: ecs-agent/daemonimages/csidriver/driver/node.go (hunk 12 above)
Comment (saurabhc123) [reply ecs-agent/daemonimages/csidriver/driver/node.go:194]: @singholt - Windows takes a bit more time to provision the volumes and times out for a few scenarios. We wanted to profile the code a bit for the first few weeks. We plan to remove these or convert them to debug statements after a few releases.

Comment (singholt) [reply ecs-agent/daemonimages/csidriver/driver/node.go]: sounds good.

File: agent/engine/task_manager.go (hunk 6 above)
Comment (jiuchoe4) [reply agent/engine/task_manager.go:1621]: Updated this in [92b4941](https://github.com/aws/amazon-ecs-agent/pull/4326/commits/92b4941fe4a9b7eb6480f6eb9cf4c18565dd99d9)

File: agent/engine/task_manager_linux.go (hunk 7 above)
Comment (jiuchoe4) [reply agent/engine/task_manager_linux.go:1]: Added this in [92b4941](https://github.com/aws/amazon-ecs-agent/pull/4326/commits/92b4941fe4a9b7eb6480f6eb9cf4c18565dd99d9)

File: ecs-agent/daemonimages/csidriver/driver/node.go (hunk 8 above)
Comment (jiuchoe4) [reply ecs-agent/daemonimages/csidriver/driver/node.go]: These logs were used for profiling Windows execution times. They are no longer needed, and as such I removed the logs in `node.go` in [92b4941](https://github.com/aws/amazon-ecs-agent/pull/4326/commits/92b4941fe4a9b7eb6480f6eb9cf4c18565dd99d9)

File: ecs-agent/daemonimages/csidriver/driver/node.go (hunk 9 above)
Comment (jiuchoe4) [reply ecs-agent/daemonimages/csidriver/driver/node.go]: Fixed in [502174e](https://github.com/aws/amazon-ecs-agent/pull/4326/commits/502174ee6c28610cf856718ec299d900f98eb65b)

File: ecs-agent/daemonimages/csidriver/driver/node.go (hunk 11 above)
Comment (jiuchoe4) [reply ecs-agent/daemonimages/csidriver/driver/node.go]: These have been removed in [502174e](https://github.com/aws/amazon-ecs-agent/pull/4326/commits/502174ee6c28610cf856718ec299d900f98eb65b)

File: ecs-agent/daemonimages/csidriver/driver/node.go (hunk 13 above)
Comment (jiuchoe4) [reply ecs-agent/daemonimages/csidriver/driver/node.go]: Removed in [502174e](https://github.com/aws/amazon-ecs-agent/pull/4326/commits/502174ee6c28610cf856718ec299d900f98eb65b)

**Review by singholt — APPROVED** [review]

**Review by mye956 — APPROVED** [review]

**Pull request merged**